File: app/Shareify Code/Shareify Code/Models/ServerManager.swift
import Foundation

class ServerManager {
    static let shared = ServerManager()
    private var cryptoManager: CryptoManager?
    private var sessionEstablished = false
    
    private init() {
        let clientId = UserDefaults.standard.string(forKey: "client_id") ?? UUID().uuidString
        UserDefaults.standard.set(clientId, forKey: "client_id")
        self.cryptoManager = CryptoManager(clientId: clientId)
    }
    
    func establishEncryptedSession(completion: @escaping (Bool) -> Void) {
        guard let publicKeyPEM = cryptoManager?.getPublicKeyPEM(),
              let jwtToken = UserDefaults.standard.string(forKey: "jwt_token") else {
            print("[Encryption] Missing public key or JWT token")
            completion(false)
            return
        }
        
        guard let url = URL(string: "https://bridge.bbarni.hackclub.app/cloud/establish_session") else {
            completion(false)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(jwtToken)", forHTTPHeaderField: "Authorization")
        
        let clientId = UserDefaults.standard.string(forKey: "client_id") ?? ""
        let requestBody: [String: Any] = [
            "client_id": clientId,
            "public_key": publicKeyPEM
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: requestBody)
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else {
                completion(false)
                return
            }
            
            if let error = error {
                print("[Encryption] Session establishment error: \(error)")
                completion(false)
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let encryptedKey = json["encrypted_session_key"] as? String else {
                print("[Encryption] Invalid response from server")
                completion(false)
                return
            }
            
            if self.cryptoManager?.decryptSessionKey(encryptedKeyBase64: encryptedKey) == true {
                self.sessionEstablished = true
                print("[Encryption] E2E encryption established successfully")
                completion(true)
            } else {
                print("[Encryption] Failed to decrypt session key")
                completion(false)
            }
        }.resume()
    }
    
    func executeServerCommand(command: String, method: String = "GET", body: [String: Any] = [:], waitTime: Int = 2, completion: @escaping (Result<Any, Error>) -> Void) {
        executeServerCommand(command: command, method: method, body: body, waitTime: waitTime, useEncryption: true, completion: completion)
    }
    
    func executeServerCommand(command: String, method: String = "GET", body: [String: Any] = [:], waitTime: Int = 2, useEncryption: Bool = true, completion: @escaping (Result<Any, Error>) -> Void) {
        
        if useEncryption && !sessionEstablished {
            establishEncryptedSession { [weak self] success in
                if success {
                    self?.executeServerCommand(command: command, method: method, body: body, waitTime: waitTime, useEncryption: useEncryption, completion: completion)
                } else {
                    print("[Encryption] Failed to establish session, falling back to unencrypted")
                    self?.executeServerCommand(command: command, method: method, body: body, waitTime: waitTime, useEncryption: false, completion: completion)
                }
            }
            return
        }
        
        guard let jwtToken = UserDefaults.standard.string(forKey: "jwt_token"), !jwtToken.isEmpty else {
            completion(.failure(ServerError.noJWTToken))
            return
        }
        
        guard let url = URL(string: "https://command.bbarni.hackclub.app/") else {
            completion(.failure(ServerError.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(jwtToken)", forHTTPHeaderField: "Authorization")

        if let shareifyJWT = UserDefaults.standard.string(forKey: "shareify_jwt"), !shareifyJWT.isEmpty {
            request.setValue(shareifyJWT, forHTTPHeaderField: "X-Shareify-JWT")
        }

        var requestBody: [String: Any] = [
            "command": command,
            "method": method,
            "wait_time": waitTime
        ]

        if !body.isEmpty {
            requestBody["body"] = body
        }
        
        let clientId = UserDefaults.standard.string(forKey: "client_id") ?? ""
        
        if useEncryption, let encryptedPayload = cryptoManager?.encryptRequest(requestBody) {
            requestBody = [
                "encrypted": true,
                "client_id": clientId,
                "encrypted_payload": encryptedPayload
            ]
            print("[Encryption] Sending encrypted command")
        } else if useEncryption {
            print("[Encryption] Failed to encrypt, sending unencrypted")
        }

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                
                if let error = error {
                    completion(.failure(error))
                    return
                }
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    completion(.failure(ServerError.invalidResponse))
                    return
                }
                
                guard let data = data else {
                    completion(.failure(ServerError.noData))
                    return
                }

                if String(data: data, encoding: .utf8) != nil {
                    print("[ServerManager] command=\(command) status=\(httpResponse.statusCode)")
                }
                
                if httpResponse.statusCode == 401 {
                    if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                       let errorMessage = json["error"] as? String,
                       errorMessage == "Invalid or expired JWT token" {
                        self.refreshJWTTokenAndRetry(originalCommand: command, originalMethod: method, originalBody: body, originalWaitTime: waitTime, useEncryption: useEncryption, completion: completion)
                        return
                    }
                }
                
                do {
                    let json = try JSONSerialization.jsonObject(with: data)
                    
                    if let jsonDict = json as? [String: Any] {
                        if useEncryption, let encrypted = jsonDict["encrypted"] as? Bool, encrypted,
                           let encryptedResponse = jsonDict["encrypted_response"] as? [String: String],
                           let decryptedJson = self.cryptoManager?.decryptResponse(encryptedPackage: encryptedResponse) {
                            print("[Encryption] Received and decrypted response")
                            completion(.success(decryptedJson))
                            return
                        }
                        
                        if let errorMessage = jsonDict["error"] as? String {
                            self.handleServerError(errorMessage: errorMessage, originalCommand: command, originalMethod: method, originalBody: body, originalWaitTime: waitTime, useEncryption: useEncryption, completion: completion)
                            return
                        }
                        
                        if let success = jsonDict["success"] as? Bool, !success {
                            let errorMessage = jsonDict["error"] as? String ?? "Unknown server error"
                            self.handleServerError(errorMessage: errorMessage, originalCommand: command, originalMethod: method, originalBody: body, originalWaitTime: waitTime, useEncryption: useEncryption, completion: completion)
                            return
                        }
                        
                        if httpResponse.statusCode == 200 {
                            completion(.success(jsonDict))
                        } else {
                            let errorMessage = jsonDict["error"] as? String ?? "Unknown server error"
                            self.handleServerError(errorMessage: errorMessage, originalCommand: command, originalMethod: method, originalBody: body, originalWaitTime: waitTime, useEncryption: useEncryption, completion: completion)
                        }
                    } else if json is [Any] {
                        if httpResponse.statusCode == 200 {
                            completion(.success(json))
                        } else {
                            self.handleServerError(errorMessage: "HTTP \(httpResponse.statusCode)", originalCommand: command, originalMethod: method, originalBody: body, originalWaitTime: waitTime, useEncryption: useEncryption, completion: completion)
                        }
                    } else {
                        completion(.failure(ServerError.invalidJSONResponse))
                    }
                } catch {
                    completion(.failure(error))
                }
            }
        }.resume()
    }
    
    private func handleServerError(errorMessage: String, originalCommand: String, originalMethod: String, originalBody: [String: Any], originalWaitTime: Int = 2, useEncryption: Bool = true, completion: @escaping (Result<Any, Error>) -> Void) {
        let isAuthError = errorMessage.lowercased().contains("unauthorized") || 
                         errorMessage.lowercased().contains("token") || 
                         errorMessage.lowercased().contains("auth") || 
                         errorMessage == "Unauthorized" || 
                         errorMessage == "Invalid token"
        
        if isAuthError {
            if let username = UserDefaults.standard.string(forKey: "server_username"),
               let password = UserDefaults.standard.string(forKey: "server_password"),
               !username.isEmpty, !password.isEmpty {
                
                loginToServer(username: username, password: password) { result in
                    switch result {
                    case .success(_):
                        self.executeServerCommand(command: originalCommand, method: originalMethod, body: originalBody, waitTime: originalWaitTime, useEncryption: useEncryption, completion: completion)
                    case .failure(let error):
                        if let serverError = error as? ServerError,
                           case .serverError(let message) = serverError,
                           message.lowercased().contains("unauthorized") {
                            NotificationCenter.default.post(name: NSNotification.Name("RedirectToLogin"), object: nil)
                        } else {
                            NotificationCenter.default.post(name: NSNotification.Name("ShowServerError"), object: nil)
                        }
                        completion(.failure(error))
                    }
                }
            } else {
                completion(.failure(ServerError.serverError(errorMessage)))
            }
        } else {
            NotificationCenter.default.post(name: NSNotification.Name("ShowServerError"), object: nil)
            completion(.failure(ServerError.serverError(errorMessage)))
        }
    }
    
    func loginToServer(username: String, password: String, completion: @escaping (Result<[String: Any], Error>) -> Void) {
        guard let jwtToken = UserDefaults.standard.string(forKey: "jwt_token"), !jwtToken.isEmpty else {
            completion(.failure(ServerError.noJWTToken))
            return
        }
        
        guard let url = URL(string: "https://command.bbarni.hackclub.app/") else {
            completion(.failure(ServerError.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(jwtToken)", forHTTPHeaderField: "Authorization")
        
        let requestBody: [String: Any] = [
            "command": "/user/login",
            "method": "POST",
            "wait_time": 5,
            "body": [
                "username": username,
                "password": password
            ]
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    completion(.failure(error))
                    return
                }
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    completion(.failure(ServerError.invalidResponse))
                    return
                }
                
                guard let data = data else {
                    completion(.failure(ServerError.noData))
                    return
                }

                if let raw = String(data: data, encoding: .utf8) {
                    print("[ServerManager] /user/login response status=\(httpResponse.statusCode) -> \(raw)")
                }
                
                do {
                    let json = try JSONSerialization.jsonObject(with: data)
                    
                    if let jsonDict = json as? [String: Any] {
                        if let errorMessage = jsonDict["error"] as? String {
                            completion(.failure(ServerError.serverError(errorMessage)))
                            return
                        }
                        
                        if httpResponse.statusCode == 200 {
                            if let token = jsonDict["token"] as? String {
                                UserDefaults.standard.set(token, forKey: "shareify_jwt")
                                UserDefaults.standard.synchronize()
                                NotificationCenter.default.post(name: NSNotification.Name("ServerLoginStatusChanged"), object: nil)
                            }
                            completion(.success(jsonDict))
                        } else {
                            let errorMessage = jsonDict["error"] as? String ?? "Login failed"
                            completion(.failure(ServerError.serverError(errorMessage)))
                        }
                    } else {
                        completion(.failure(ServerError.invalidJSONResponse))
                    }
                } catch {
                    completion(.failure(error))
                }
            }
        }.resume()
    }
    
    func bridgeLogin(email: String, password: String, completion: @escaping (Result<[String: Any], Error>) -> Void) {
        guard let url = URL(string: "https://bridge.bbarni.hackclub.app/login") else {
            completion(.failure(ServerError.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let requestBody: [String: Any] = [
            "email": email,
            "password": password
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    completion(.failure(error))
                    return
                }
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    completion(.failure(ServerError.invalidResponse))
                    return
                }
                
                guard let data = data else {
                    completion(.failure(ServerError.noData))
                    return
                }
                
                if let raw = String(data: data, encoding: .utf8) {
                    print("[ServerManager] bridge login response status=\(httpResponse.statusCode) -> \(raw)")
                }
                
                do {
                    let json = try JSONSerialization.jsonObject(with: data)
                    
                    if let jsonDict = json as? [String: Any] {
                        if let errorMessage = jsonDict["error"] as? String {
                            completion(.failure(ServerError.serverError(errorMessage)))
                            return
                        }
                        
                        if httpResponse.statusCode == 200 {
                            if let jwtToken = jsonDict["jwt_token"] as? String {
                                UserDefaults.standard.set(jwtToken, forKey: "jwt_token")
                                UserDefaults.standard.set(email, forKey: "user_email")
                                UserDefaults.standard.set(password, forKey: "user_password")
                                UserDefaults.standard.synchronize()
                                NotificationCenter.default.post(name: NSNotification.Name("ServerLoginStatusChanged"), object: nil)
                            }
                            completion(.success(jsonDict))
                        } else {
                            let errorMessage = jsonDict["error"] as? String ?? "Login failed"
                            completion(.failure(ServerError.serverError(errorMessage)))
                        }
                    } else {
                        completion(.failure(ServerError.invalidJSONResponse))
                    }
                } catch {
                    completion(.failure(error))
                }
            }
        }.resume()
    }
    
    private func refreshJWTTokenAndRetry(originalCommand: String, originalMethod: String, originalBody: [String: Any], originalWaitTime: Int = 2, useEncryption: Bool = true, completion: @escaping (Result<Any, Error>) -> Void) {
        guard let email = UserDefaults.standard.string(forKey: "user_email"),
              let password = UserDefaults.standard.string(forKey: "user_password"),
              !email.isEmpty, !password.isEmpty else {
            completion(.failure(ServerError.serverError("Login credentials not available")))
            return
        }
        
        guard let url = URL(string: "https://bridge.bbarni.hackclub.app/login") else {
            completion(.failure(ServerError.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let requestBody: [String: Any] = [
            "email": email,
            "password": password
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                guard let data = data,
                      let httpResponse = response as? HTTPURLResponse,
                      httpResponse.statusCode == 200,
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let newJwtToken = json["jwt_token"] as? String else {
                    completion(.failure(ServerError.serverError("Failed to refresh login")))
                    return
                }

                if String(data: data, encoding: .utf8) != nil {
                    print("[ServerManager] bridge login refresh -> success")
                }
                
                UserDefaults.standard.set(newJwtToken, forKey: "jwt_token")
                UserDefaults.standard.synchronize()
                
                self.sessionEstablished = false
                
                self.executeServerCommand(command: originalCommand, method: originalMethod, body: originalBody, waitTime: originalWaitTime, useEncryption: useEncryption, completion: completion)
            }
        }.resume()
    }
    
    func isServerLoggedIn() -> Bool {
        let hasShareifyJWT = UserDefaults.standard.string(forKey: "shareify_jwt") != nil
        return hasShareifyJWT
    }
    
    func testServerConnection(completion: @escaping (Bool) -> Void) {
        guard let jwtToken = UserDefaults.standard.string(forKey: "jwt_token"), !jwtToken.isEmpty else {
            completion(false)
            return
        }
        
        guard let url = URL(string: "https://command.bbarni.hackclub.app/") else {
            completion(false)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(jwtToken)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 3
        
        if let shareifyJWT = UserDefaults.standard.string(forKey: "shareify_jwt"), !shareifyJWT.isEmpty {
            request.setValue(shareifyJWT, forHTTPHeaderField: "X-Shareify-JWT")
        }
        
        let requestBody: [String: Any] = [
            "command": "/is_up",
            "method": "GET",
            "wait_time": 1
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        } catch {
            completion(false)
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let _ = error {
                DispatchQueue.main.async {
                    completion(false)
                }
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                DispatchQueue.main.async {
                    completion(httpResponse.statusCode == 200 || httpResponse.statusCode == 404)
                }
            } else {
                DispatchQueue.main.async {
                    completion(false)
                }
            }
        }.resume()
    }
}

enum ServerError: LocalizedError {
    case noJWTToken
    case invalidURL
    case invalidResponse
    case noData
    case invalidJSONResponse
    case serverError(String)
    case encryptionSetupFailed
    
    var errorDescription: String? {
        switch self {
        case .noJWTToken:
            return "JWT token not found"
        case .invalidURL:
            return "Invalid server URL"
        case .invalidResponse:
            return "Invalid response from server"
        case .noData:
            return "No data received from server"
        case .invalidJSONResponse:
            return "Invalid JSON response"
        case .serverError(let message):
            return message
        case .encryptionSetupFailed:
            return "Failed to establish encrypted connection"
        }
    }
}

File: app/Shareify Code Android/app/src/main/java/com/shareify/code/models/ServerManager.kt
package com.shareify.code.models

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.TimeUnit

class ServerManager private constructor(private val context: Context) {
    private var cryptoManager: CryptoManager? = null
    private var sessionEstablished = false
    private val prefs: SharedPreferences = context.getSharedPreferences("server_prefs", Context.MODE_PRIVATE)
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    init {
        val clientId = prefs.getString("client_id", null) ?: UUID.randomUUID().toString()
        prefs.edit().putString("client_id", clientId).apply()
        cryptoManager = CryptoManager(clientId)
    }

    suspend fun establishEncryptedSession(): Boolean = withContext(Dispatchers.IO) {
        val publicKeyPEM = cryptoManager?.getPublicKeyPEM() ?: return@withContext false
        val jwtToken = prefs.getString("jwt_token", null) ?: return@withContext false

        try {
            val clientId = prefs.getString("client_id", "") ?: ""
            val requestBody = JSONObject().apply {
                put("client_id", clientId)
                put("public_key", publicKeyPEM)
            }

            val request = Request.Builder()
                .url("https://bridge.bbarni.hackclub.app/cloud/establish_session")
                .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                .addHeader("Authorization", "Bearer $jwtToken")
                .build()

            val response = client.newCall(request).execute()
            val responseBody = response.body?.string()

            if (response.isSuccessful && responseBody != null) {
                val json = JSONObject(responseBody)
                val encryptedKey = json.getString("encrypted_session_key")
                
                sessionEstablished = cryptoManager?.decryptSessionKey(encryptedKey) ?: false
                sessionEstablished
            } else {
                false
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    suspend fun executeServerCommand(
        command: String,
        method: String = "GET",
        body: Map<String, Any> = emptyMap(),
        waitTime: Int = 2,
        useEncryption: Boolean = true
    ): Result<Any> = withContext(Dispatchers.IO) {
        if (useEncryption && !sessionEstablished) {
            val established = establishEncryptedSession()
            if (!established) {
                return@withContext executeServerCommand(command, method, body, waitTime, false)
            }
        }

        val jwtToken = prefs.getString("jwt_token", null)
        if (jwtToken.isNullOrEmpty()) {
            return@withContext Result.failure(ServerError.NoJWTToken)
        }

        try {
            var requestBody = JSONObject().apply {
                put("command", command)
                put("method", method)
                put("wait_time", waitTime)
                if (body.isNotEmpty()) {
                    put("body", JSONObject(body))
                }
            }

            val clientId = prefs.getString("client_id", "") ?: ""
            
            if (useEncryption) {
                val encryptedPayload = cryptoManager?.encryptRequest(requestBody.toString())
                if (encryptedPayload != null) {
                    requestBody = JSONObject().apply {
                        put("encrypted", true)
                        put("client_id", clientId)
                        put("encrypted_payload", encryptedPayload)
                    }
                }
            }

            val requestBuilder = Request.Builder()
                .url("https://command.bbarni.hackclub.app/")
                .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                .addHeader("Authorization", "Bearer $jwtToken")

            val shareifyJWT = prefs.getString("shareify_jwt", null)
            if (!shareifyJWT.isNullOrEmpty()) {
                requestBuilder.addHeader("X-Shareify-JWT", shareifyJWT)
            }

            val response = client.newCall(requestBuilder.build()).execute()
            val responseBody = response.body?.string()

            if (response.code == 401) {
                return@withContext refreshJWTAndRetry(command, method, body, waitTime, useEncryption)
            }

            if (response.isSuccessful && responseBody != null) {
                try {
                    val json = JSONObject(responseBody)
                    
                    if (useEncryption && json.optBoolean("encrypted", false)) {
                        val encryptedResponse = json.getJSONObject("encrypted_response")
                        val decrypted = cryptoManager?.decryptResponse(encryptedResponse)
                        if (decrypted != null) {
                            return@withContext Result.success(decrypted)
                        }
                    }
                    
                    Result.success(json.toMap())
                } catch (e: Exception) {
                    try {
                        val jsonArray = JSONArray(responseBody)
                        Result.success(jsonArray.toList())
                    } catch (e2: Exception) {
                        Result.failure(ServerError.InvalidResponse)
                    }
                }
            } else {
                Result.failure(ServerError.ServerError(responseBody ?: "Unknown error"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private suspend fun refreshJWTAndRetry(
        command: String,
        method: String,
        body: Map<String, Any>,
        waitTime: Int,
        useEncryption: Boolean
    ): Result<Any> = withContext(Dispatchers.IO) {
        val email = prefs.getString("user_email", null)
        val password = prefs.getString("user_password", null)

        if (email.isNullOrEmpty() || password.isNullOrEmpty()) {
            return@withContext Result.failure(ServerError.NoCredentials)
        }

        try {
            val requestBody = JSONObject().apply {
                put("email", email)
                put("password", password)
            }

            val request = Request.Builder()
                .url("https://bridge.bbarni.hackclub.app/login")
                .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                .build()

            val response = client.newCall(request).execute()
            val responseBody = response.body?.string()

            if (response.isSuccessful && responseBody != null) {
                val json = JSONObject(responseBody)
                val newJwtToken = json.getString("jwt_token")
                prefs.edit().putString("jwt_token", newJwtToken).apply()
                
                sessionEstablished = false
                
                executeServerCommand(command, method, body, waitTime, useEncryption)
            } else {
                Result.failure(ServerError.AuthFailed)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun loginToServer(username: String, password: String): Result<Map<String, Any>> = 
        withContext(Dispatchers.IO) {
            val jwtToken = prefs.getString("jwt_token", null)
            if (jwtToken.isNullOrEmpty()) {
                return@withContext Result.failure(ServerError.NoJWTToken)
            }

            try {
                val requestBody = JSONObject().apply {
                    put("command", "/user/login")
                    put("method", "POST")
                    put("wait_time", 5)
                    put("body", JSONObject().apply {
                        put("username", username)
                        put("password", password)
                    })
                }

                val request = Request.Builder()
                    .url("https://command.bbarni.hackclub.app/")
                    .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                    .addHeader("Authorization", "Bearer $jwtToken")
                    .build()

                val response = client.newCall(request).execute()
                val responseBody = response.body?.string()

                if (response.isSuccessful && responseBody != null) {
                    val json = JSONObject(responseBody)
                    val token = json.optString("token")
                    if (token.isNotEmpty()) {
                        prefs.edit()
                            .putString("shareify_jwt", token)
                            .putString("server_username", username)
                            .putString("server_password", password)
                            .apply()
                    }
                    Result.success(json.toMap())
                } else {
                    Result.failure(ServerError.ServerError(responseBody ?: "Login failed"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    suspend fun bridgeLogin(email: String, password: String): Result<Map<String, Any>> = 
        withContext(Dispatchers.IO) {
            try {
                val requestBody = JSONObject().apply {
                    put("email", email)
                    put("password", password)
                }

                val request = Request.Builder()
                    .url("https://bridge.bbarni.hackclub.app/login")
                    .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                    .build()

                val response = client.newCall(request).execute()
                val responseBody = response.body?.string()

                if (response.isSuccessful && responseBody != null) {
                    val json = JSONObject(responseBody)
                    val jwtToken = json.getString("jwt_token")
                    prefs.edit()
                        .putString("jwt_token", jwtToken)
                        .putString("user_email", email)
                        .putString("user_password", password)
                        .apply()
                    Result.success(json.toMap())
                } else {
                    Result.failure(ServerError.ServerError(responseBody ?: "Bridge login failed"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    fun isServerLoggedIn(): Boolean {
        return !prefs.getString("shareify_jwt", null).isNullOrEmpty()
    }

    suspend fun testServerConnection(): Boolean = withContext(Dispatchers.IO) {
        val jwtToken = prefs.getString("jwt_token", null)
        if (jwtToken.isNullOrEmpty()) return@withContext false

        try {
            val requestBody = JSONObject().apply {
                put("command", "/is_up")
                put("method", "GET")
                put("wait_time", 1)
            }

            val requestBuilder = Request.Builder()
                .url("https://command.bbarni.hackclub.app/")
                .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
                .addHeader("Authorization", "Bearer $jwtToken")

            val shareifyJWT = prefs.getString("shareify_jwt", null)
            if (!shareifyJWT.isNullOrEmpty()) {
                requestBuilder.addHeader("X-Shareify-JWT", shareifyJWT)
            }

            val response = client.newCall(requestBuilder.build()).execute()
            response.isSuccessful || response.code == 404
        } catch (e: Exception) {
            false
        }
    }

    companion object {
        @Volatile
        private var instance: ServerManager? = null

        fun getInstance(context: Context): ServerManager {
            return instance ?: synchronized(this) {
                instance ?: ServerManager(context.applicationContext).also { instance = it }
            }
        }
    }
}

sealed class ServerError : Exception() {
    object NoJWTToken : ServerError()
    object NoCredentials : ServerError()
    object InvalidResponse : ServerError()
    object AuthFailed : ServerError()
    data class ServerError(val message: String) : com.shareify.code.models.ServerError()
}

private fun JSONObject.toMap(): Map<String, Any> {
    val map = mutableMapOf<String, Any>()
    keys().forEach { key ->
        map[key] = get(key)
    }
    return map
}

private fun JSONArray.toList(): List<Any> {
    val list = mutableListOf<Any>()
    for (i in 0 until length()) {
        list.add(get(i))
    }
    return list
}
